apps/emqx/src/emqx_persistent_session_ds_state.erl

```diff
@@ -478,94 +593,221 @@ pmap_del(
 pmap_fold(Fun, Acc, #pmap{cache = Cache}) ->
     maps:fold(Fun, Acc, Cache).
 
--spec pmap_commit(emqx_persistent_session_ds:id(), pmap(K, V)) -> pmap(K, V).
+-spec pmap_commit(emqx_persistent_session_ds:id(), pmap(K, V)) ->
+    {{[emqx_types:message()], [term()]}, pmap(K, V)}.
 pmap_commit(
-    SessionId, Pmap = #pmap{table = Tab, dirty = Dirty, cache = Cache}
+    SessionId, Pmap = #pmap{domain = Domain, dirty = Dirty, cache = Cache}
 ) ->
-    maps:foreach(
-        fun
-            (K, del) ->
-                kv_pmap_delete(Tab, SessionId, K);
-            (K, dirty) ->
-                V = maps:get(K, Cache),
-                kv_pmap_persist(Tab, SessionId, K, V)
-        end,
-        Dirty
-    ),
-    Pmap#pmap{
+    Out =
+        maps:fold(
+            fun
+                (K, del, {AccPersist, AccDel}) ->
+                    {AccPersist, [K | AccDel]};
+                (K, dirty, {AccPersist, AccDel}) ->
+                    V = maps:get(K, Cache),
+                    Msg = to_domain_msg(Domain, SessionId, K, V),
+                    {[Msg | AccPersist], AccDel}
+            end,
+            {[], []},
+            Dirty
+        ),
+    {Out, Pmap#pmap{
         dirty = #{}
-    }.
+    }}.
 
 -spec pmap_format(pmap(_K, _V)) -> map().
 pmap_format(#pmap{cache = Cache}) ->
     Cache.
 
 %% Functions dealing with set tables:
 
-kv_persist(Tab, SessionId, Val0) ->
-    Val = encoder(encode, Tab, Val0),
-    mnesia:write(Tab, #kv{k = SessionId, v = Val}, write).
+session_restore(SessionId) ->
+    Empty = maps:from_keys(
+        [
+            ?metadata_domain,
+            ?subscription_domain,
+            ?stream_domain,
+            ?rank_domain,
+            ?seqno_domain
+        ],
+        []
+    ),
+    Data = maps:groups_from_list(
+        fun(#{domain := Domain}) ->
+            Domain
+        end,
+        read_iterate(SessionId, ['#'])
+    ),
+    maps:merge(Empty, Data).
 
-kv_restore(Tab, SessionId) ->
-    [encoder(decode, Tab, V) || #kv{v = V} <- mnesia:read(Tab, SessionId)].
+%%
 
-%% Functions dealing with bags:
+to_domain_msg(Domain, SessionId, Key, Val) ->
+    #message{
+        %% unused; empty binary to satisfy dialyzer
+        id = <<>>,
+        timestamp = 0,
+        from = SessionId,
+        topic = domain_topic_encode(Domain, SessionId, Key),
+        payload = val_encode(Domain, Val)
+    }.
 
-%% @doc Create a mnesia table for the PMAP:
--spec create_kv_pmap_table(atom()) -> ok.
-create_kv_pmap_table(Table) ->
-    mria:create_table(Table, [
-        {type, ordered_set},
-        {rlog_shard, ?DS_MRIA_SHARD},
-        {storage, rocksdb_copies},
-        {record_name, kv},
-        {attributes, record_info(fields, kv)}
+from_domain_msg(#message{topic = Topic, payload = Bin}) ->
+    #{
+        domain := Domain,
+        session_id := _SessionId,
+        key := _Key
+    } = Data = domain_topic_decode(Topic),
+    Data#{val => val_decode(Domain, Bin)}.
+
+-spec domain_topic_encode
```

Comment:
I find that a 30-line signature to a 2-line function that just passes the parameters into another function significantly *reduces* readability of the code. This is an internal function within a self-contained module that handles a well-defined set of data.

Reply:
Removed.